homeassistant/components/tradfri/__init__.py

```diff
@@ -139,6 +143,69 @@ async def async_keep_alive(now: datetime) -> None:
     return True
 
 
+def migrate_config_entry_and_identifiers(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> None:
+    """Migrate old non-unique identifiers to new unique identifiers."""
+
+    related_device_flag: bool
+    device_id: str
+
+    device_reg = dr.async_get(hass)
+    # pylint:disable=R1702
+    # Get all devices associated to contextual gateway config_entry and loop through list of devices.
+    for device in dr.async_entries_for_config_entry(device_reg, config_entry.entry_id):
+        related_device_flag = False
+        for identifier in device.identifiers:
+            if identifier[0] != DOMAIN:
+                continue
+
+            related_device_flag = True
+
+            _id = identifier[1]
+
+            # Identify gateway device.
+            if _id == config_entry.data[CONF_GATEWAY_ID]:
+                related_device_flag = False  # Using this to avoid updating gateway's own device registry entry
+                break
+
+            device_id = str(_id)
+            break
+
+        # Check that device is related to tradfri domain (and is not the gateway itself)
+        if related_device_flag:
+            # Identify where a device has multiple config_entry IDs
+            config_entry_ids = device.config_entries
+            if len(config_entry_ids) > 1:
+                # Loop through list of config_entry_ids for device
+                for config_entry_id in config_entry_ids:
+                    # Check that the config entry in list is not the device's primary config entry
+                    if config_entry_id != device.primary_config_entry:
+                        # Check that the 'other' config entry is also a tradfri config entry
+                        other_entry = hass.config_entries.async_get_entry(
+                            config_entry_id
+                        )
+                        # mypy linter 'union-attr' error occurs here without exclusion -
+                        # other_entry could be None (and therefore won't have a domain), but that would
+                        # only happen if config entries were manually changed in core.config_entries.
+                        if other_entry.domain == DOMAIN:
+                            # Remove non-primary 'tradfri' config entry from device's config_entry_ids
+                            device_reg.async_update_device(
+                                device.id, remove_config_entry_id=config_entry_id
+                            )
+
+            if device_id.find(config_entry.data[CONF_GATEWAY_ID]) == -1:
+                device_reg.async_update_device(
+                    device.id,
+                    new_identifiers={
+                        (DOMAIN, f"{device_id}-{config_entry.data[CONF_GATEWAY_ID]}")
```

Comment:
I'd start with the gateway id so that the whole string is the same as the unique id of the entity.